src/main/scala/scala/concurrent/java8/FutureConvertersImpl.scala

```diff
@@ -66,13 +66,20 @@ object FuturesConvertersImpl {
       cf
     }
 
-    override def toCompletableFuture(): CompletableFuture[T] =
-      throw new UnsupportedOperationException("this CompletionStage represents a read-only Scala Future")
+    /**
+     * @inheritdoc
+     *
+     * WARNING: completing the result of this method will not complete the underlying
+     *          Scala Future or Promise.
+     */
+    override def toCompletableFuture(): CompletableFuture[T] = {
+      this // TODO or maybe `thenApply(JF.identity())`
+    }
 
     override def toString: String = super[CompletableFuture].toString
   }
 
-  class P[T] extends impl.Promise.DefaultPromise[T] with BiConsumer[T, Throwable] {
+  class P[T](val wrapped: CompletionStage[T]) extends impl.Promise.DefaultPromise[T] with BiConsumer[T, Throwable] {
```

Comment:
The only place that creates instance of `P` would not be reached with `cs : CF`:

```scala
   def toScala[T](cs: CompletionStage[T]): Future[T] = {
    cs match {
      case cf: CF[T] => cf.wrapped
      case _ =>
        val p = new P[T](cs)
        cs whenComplete p
        p.future
    }
  }
```

Do you disagree? What problems do you envisiage if wrapped *was* a `CF`?